### DISCUSSION TITLE
UUID shapes?

### DISCUSSION THREAD

**Discussion by Quafadas** [disneystreaming/smithy4s#262] [Q&A] [answered]
In my continuing quest to dominate the smithy discussion boards... 

```
structure EventId {
  @httpLabel
  @required
  name:UUID
}
```
Gets me this; 
```
member shape targets an unresolved shape `smithy4s.hello#UUID`
```
I guess I was hoping that UUID was somehow built in. Is the canonical strategy here to have a string, validate on receipt and return a "NotUuuid" error if the string doesn't comply?

**Comment by Baccata** [comment, marked as answer]
> I guess I was hoping that UUID was somehow built in. Is the canonical strategy here to have a string,  

Well you're in luck, because we make use of UUID types heavily internally so we decided to add first class support for it. 

``` 
use smithy4s.api#UUID

structure EventId {
  @httpLabel
  @required
  name:UUID
}
```

the `name` field gets rendered as a `java.util.UUID`. If you want newtype UUIDs, you can do : 

```  
use smithy4s.api#uuidFormat

@uuidFormat 
string MyUUID 
```

**Reply by Quafadas** [reply]
That is... excellent. Thanks!

**Comment by Baccata** [comment]
> Is the canonical strategy here to have a string, validate on receipt and return a "NotUuuid" error if the string doesn't comply 

I'm not sure there is such a thing as the "canonical strategy". Maybe it's a case that AWS uses another format that UUID and didn't deem there was a justification for UUID to become a bonafide shape 🤷 . Maybe it's a case that validating whether something is a UUID is not that important for most applications, and that the value of UUIDs derives from their random generation as opposed to the format. 

Without a custom `uuidFormat`, I've seen people use the `@pattern` trait to annotate a string with a regex. Smithy4s understands `@pattern` and performs validation against it where relevant

**Reply by Quafadas** [reply]
Thankyou again for the time and all the extra colour!
